### PR TITLE
Feature: Federation duplicated ontologies canonical portal

### PR DESCRIPTION
**Pull request by Bilelkihal**
### PR description
This pull request introduces a function to determine the canonical portal for duplicated ontologies within a federation based on the :pullLocation data. The canonical portal is identified as the one that appears most frequently across the pullLocations of different duplicates.

### Changes
- Added a function to retrieve the canonical portal for duplicated ontologies using :pullLocation.
- Open the ontologies in their canonical portal when they are duplicated.
- Added a function to retrieve the canonical portal for duplicated search page results using :pullLocation of their submissions.
- Open the search results in their canonical portal when they are duplicated.

ps. for the ontologies present in the current portal, they open always in the current portal.

### CODE REVIEW

File: app/helpers/federation_helper.rb [hunk 1]
```diff
@@ -124,4 +124,25 @@ def federation_external_class?(class_object)
   end
 
 
+  def canonical_portal(ontologies)
+    portals = federated_portals.map{|portal| portal.first}
```
Comment (syphax-bouazzouni) [inline app/helpers/federation_helper.rb:128]: can be replaced with `federated_portals.keys`

File: app/helpers/federation_helper.rb [hunk 2]
```diff
@@ -124,4 +124,25 @@ def federation_external_class?(class_object)
...
+    portal_counts = Hash.new(0)
+    # Count occurrences of each portal in the pull_location URL
+      portals.each do |portal|
+        if ontology[:pullLocation]&.include?(portal.to_s)
```
Comment (syphax-bouazzouni) [inline app/helpers/federation_helper.rb:132]: return an error because `ontology` is not defined.

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/controllers/concerns/submission_filter.rb [hunk 3]
```diff
@@ -71,7 +71,7 @@ def merge_by_acronym(submissions)
       if ontologies.size.eql?(1)
         ontology = ontologies.first
       else
-        ontology = ontologies.select { |x| helpers.internal_ontology?(x[:id]) }.first || ontologies.first
+        ontology = ontologies.select { |x| helpers.internal_ontology?(x[:id]) }.first || ontology_from_portal(ontologies, canonical_portal(ontologies))
```
Comment (syphax-bouazzouni) [inline app/controllers/concerns/submission_filter.rb:74]: maybe all this line can be a function called `find_canonical_ontology`

File: app/helpers/federation_helper.rb (hunk 2 above)
Comment (Bilelkihal) [reply app/helpers/federation_helper.rb]: When you did the merge of development, the line of ontologies loop changed position in a weird way. Fixed now.

File: app/controllers/concerns/submission_filter.rb [hunk 4]
```diff
@@ -80,6 +80,10 @@ def merge_by_acronym(submissions)
     merged_submissions
   end
 
+  def canonical_portal_ontology(ontologies)
```
Comment (syphax-bouazzouni) [inline app/controllers/concerns/submission_filter.rb:83]: this should go to federation helper

File: app/helpers/federation_helper.rb [hunk 5]
```diff
@@ -123,5 +123,63 @@ def federation_external_class?(class_object)
...
+    ontologies.select{|o| o[:id].include?(portal.to_s)}.first
+  end
+
+  def apply_canonical_portal(search_results, all_submissions)
+    search_results.each do |result|
+      next if result[:root][:portal_name].nil? || result[:root][:other_portals].blank?
+
+      candidates = [result[:root][:link]] + result[:root][:other_portals].map { |p| p[:link].split('?').first }
+
+      portal_counts = Hash.new(0)
```
Comment (syphax-bouazzouni) [inline app/helpers/federation_helper.rb:161]: from lines `161` to `170`  is the same as in line `139` in the function `external_canonical_ontology_portal`

File: app/controllers/concerns/search_aggregator.rb [hunk 6]
```diff
@@ -31,12 +31,16 @@ def aggregate_results(query, results)
...
 
+    all_submissions = LinkedData::Client::Models::OntologySubmission.all(include: 'pullLocation', include_views: true, display_links: false, display_context: false)
+
     search_results = grouped_results.map do |group|
       format_search_result(group, all_ontologies)
     end
 
     search_results = merge_sort_federated_results(query, search_results) if federation_enabled?
 
+    search_results = apply_canonical_portal(search_results, all_submissions)
```
Comment (syphax-bouazzouni) [inline app/controllers/concerns/search_aggregator.rb:42]: the canonical choice needs to be done in the function `merge_federated_results`  step as done for the ontologies in the browse page.

File: app/helpers/federation_helper.rb [hunk 7]
```diff
@@ -123,5 +123,63 @@ def federation_external_class?(class_object)
...
+
+      canonical_portal = portal_counts.max_by { |_, count| count }&.first
+      next if canonical_portal.nil? || result[:root][:portal_name].eql?(canonical_portal.to_s)
+
+      canonical_portal_result = result[:root][:other_portals].find { |r| r[:name] == canonical_portal.to_s }
+      swap_portal_attributes(result[:root], canonical_portal_result) if canonical_portal_result
+    end
+  end
+
+  def swap_portal_attributes(root_portal, new_portal)
```
Comment (syphax-bouazzouni) [inline app/helpers/federation_helper.rb:178]: this function can be rewritten like this:
```ruby
def swap_portal_attributes(root_portal, new_portal)
  [:link, :portal_name, :portal_color, :portal_light_color].each do |attribute|
    root_portal[attribute], new_portal[attribute] = new_portal[attribute], root_portal[attribute]
  end
end
```

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/controllers/concerns/search_aggregator.rb (hunk 6 above)
Comment (Bilelkihal) [reply app/controllers/concerns/search_aggregator.rb:42]: We need `all_submissions `value only for this function, this is why I didn't want to nest it in for 2 levels.
In all cases the logic here is way different than the one in the browse page.

File: app/helpers/federation_helper.rb (hunk 5 above)
Comment (Bilelkihal) [reply app/helpers/federation_helper.rb:161]: Yes We can extract that code in a function`count_portal_occurrences` but this abstraction will make the code less readable. 

We can do something like this:
```
  def count_portal_occurrences(items, key_extractor)
    portal_counts = Hash.new(0)
    items.each do |item|
      federated_portals.keys.each do |portal|
        portal_counts[portal] += 1 if key_extractor.call(item)&.include?(portal.to_s)
      end
    end
    portal_counts.max_by { |_, count| count }&.first
  end
```

and `key_extractor` is a lambda function.

But I see no need to complicate this code more.

File: app/helpers/federation_helper.rb (hunk 7 above)
Comment (syphax-bouazzouni) [reply app/helpers/federation_helper.rb:178]: Does this not work?

File: app/helpers/federation_helper.rb [hunk 8]
```diff
@@ -123,5 +123,63 @@ def federation_external_class?(class_object)
...
+      next if result[:root][:portal_name].nil? || result[:root][:other_portals].blank?
+
+      candidates = [result[:root][:link]] + result[:root][:other_portals].map { |p| p[:link].split('?').first }
+
+      portal_counts = Hash.new(0)
+      candidates.each do |candidate|
+        submission = all_submissions.find { |s| s.id.split('/')[0...-2].join('/') == candidate }
+
+        federated_portals.keys.each do |portal|
+          portal_counts[portal] += 1 if submission&.include?(portal.to_s)
```
Comment (syphax-bouazzouni) [inline app/helpers/federation_helper.rb:166]: Don't you use pullLocation at all for this function?

File: app/controllers/concerns/search_aggregator.rb (hunk 6 above)
Comment (syphax-bouazzouni) [reply app/controllers/concerns/search_aggregator.rb:42]: OK

File: app/helpers/federation_helper.rb (hunk 5 above)
Comment (syphax-bouazzouni) [reply app/helpers/federation_helper.rb:161]: you can also have a function that has as input ontologies IDs, and for the search case, you map the results to ontologies IDs. I did something similar here https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/778/commits/ac1f608256472b59ed7b1c24cf3c2bd51657f70e and https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/778/commits/a653fd1342cb941a8b0250e85e7c0a9a7f90819f

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

**Review by Bilelkihal — COMMENTED** [review]
✅

**Review by syphax-bouazzouni — APPROVED** [review]
Approved, did the following changes: 

- put again search canonical logic the search_aggregator file and refactor the code (https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/770/commits/55d0b716707e17451b797a7786201a99c5559a93)
- do the search canonical logic only if enabled (https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/770/commits/a86ba778d8783d03756f373dc339501fb704b38c)

**Pull request merged**